packages/react-checkbox/src/components/Checkbox/useCheckbox.tsx

```diff
@@ -0,0 +1,112 @@
+import * as React from 'react';
+import {
+  makeMergeProps,
+  resolveShorthandProps,
+  useControllableState,
+  useId,
+  useIsomorphicLayoutEffect,
+  useMergedRefs,
+} from '@fluentui/react-utilities';
+import { CheckboxProps, CheckboxShorthandProps, CheckboxState } from './Checkbox.types';
+import { Label } from '@fluentui/react-label';
+import { DefaultCheckmarkIcon, DefaultMixedIcon } from './DefaultIcons';
+
+/**
+ * Array of all shorthand properties listed in CheckboxShorthandProps
+ */
+export const checkboxShorthandProps: CheckboxShorthandProps[] = ['label', 'indicator', 'input'];
+
+const mergeProps = makeMergeProps<CheckboxState>({ deepMerge: checkboxShorthandProps });
+
+/**
+ * Create the state required to render Checkbox.
+ *
+ * The returned state can be modified with hooks such as useCheckboxStyles,
+ * before being passed to renderCheckbox.
+ *
+ * @param props - props from this instance of Checkbox
+ * @param ref - reference to root HTMLElement of Checkbox
+ * @param defaultProps - (optional) default prop values provided by the implementing type
+ */
+export const useCheckbox = (
+  props: CheckboxProps,
+  ref: React.Ref<HTMLElement>,
+  defaultProps?: CheckboxProps,
+): CheckboxState => {
+  const state = mergeProps(
+    {
+      ref,
+      id: useId('checkbox-'),
+      size: 'medium',
+      labelPosition: 'after',
+      label: {
+        as: Label,
+      },
+      indicator: {
+        as: 'div',
+      },
+      input: {
+        as: 'input',
+        type: 'checkbox',
+        children: null,
+      },
+    },
+    defaultProps && resolveShorthandProps(defaultProps, checkboxShorthandProps),
+    resolveShorthandProps(props, checkboxShorthandProps),
+  );
+
+  const [checked, setCheckedInternal] = useControllableState({
+    defaultState: props.defaultChecked,
+    state: props.checked,
+    initialState: false,
+  });
+
+  const setChecked = React.useCallback(
+    (ev: React.ChangeEvent<HTMLInputElement>, val: boolean | 'mixed') => {
+      const onChange = state.onChange;
+      onChange?.(ev, { checked: val });
+      setCheckedInternal(val);
+    },
+    [state.onChange, setCheckedInternal],
+  );
+  state.input.checked = checked === true;
+  state.checked = checked ? checked : false;
+  state.indicator.children = checked === 'mixed' ? <DefaultMixedIcon /> : <DefaultCheckmarkIcon />;
+
+  const userOnChange = state.input.onChange;
+  state.input.onChange = React.useCallback(
```

Comment:
It's more reasonable to use `useEventCallback` in this case if you would like to make it stable as if user will pass a closure, it could produce re-renders:

```ts
<Checkbox input={{ onChange: () => {} />}} />
//                           👆 reference is not stable => will recreate "state.input.onChange"
```

---

```diff
-  state.input.onChange = React.useCallback(
+  state.input.onChange = useEventCallback(
    ev => {
      userOnChange?.(ev);
      setChecked(ev, ev.currentTarget.indeterminate ? 'mixed' : ev.currentTarget.checked);
    },
-    [userOnChange, setChecked],
  );
```

`useEventCallback` will create a stable callback, https://reactjs.org/docs/hooks-faq.html#how-to-read-an-often-changing-value-from-usecallback.

Reply:
I see, thank you for the reference!